src/coreComponents/physicsSolvers/wavePropagation/AcousticWaveEquationSEM.cpp

```diff
@@ -0,0 +1,821 @@
+  {
+
+    MeshLevel & meshLevel =  meshBody.getMeshLevel( 0 );
+
+    NodeManager & nodeManager = meshLevel.getNodeManager();
+
+    nodeManager.registerExtrinsicData< extrinsicMeshData::Pressure_nm1,
+                                       extrinsicMeshData::Pressure_n,
+                                       extrinsicMeshData::Pressure_np1,
+                                       extrinsicMeshData::ForcingRHS,
+                                       extrinsicMeshData::MassVector,
+                                       extrinsicMeshData::DampingVector,
+                                       extrinsicMeshData::StiffnessVector,
+                                       extrinsicMeshData::FreeSurfaceNodeIndicator >( this->getName() );
+
+    FaceManager & faceManager = meshLevel.getFaceManager();
+    faceManager.registerExtrinsicData< extrinsicMeshData::FreeSurfaceFaceIndicator >( this->getName() );
+
+    ElementRegionManager & elemManager = meshLevel.getElemManager();
+
+    elemManager.forElementSubRegions< CellElementSubRegion >( [&]( CellElementSubRegion & subRegion )
+    {
+      subRegion.registerExtrinsicData< extrinsicMeshData::MediumVelocity >( this->getName() );
+    } );
+
+  } );
+}
+
+
+void AcousticWaveEquationSEM::precomputeSourceAndReceiverTerm( MeshLevel & mesh )
+{
+  NodeManager & nodeManager = mesh.getNodeManager();
+
+  arrayView2d< real64 const, nodes::REFERENCE_POSITION_USD > const X = nodeManager.referencePosition().toViewConst();
+
+  arrayView2d< real64 const > const sourceCoordinates = m_sourceCoordinates.toViewConst();
+  arrayView2d< localIndex > const sourceNodeIds = m_sourceNodeIds.toView();
+  arrayView2d< real64 > const sourceConstants = m_sourceConstants.toView();
+  arrayView1d< localIndex > const sourceIsLocal = m_sourceIsLocal.toView();
+  sourceNodeIds.setValues< serialPolicy >( -1 );
+  sourceConstants.setValues< serialPolicy >( -1 );
+  sourceIsLocal.setValues< serialPolicy >( 0 );
+
+  arrayView2d< real64 const > const receiverCoordinates = m_receiverCoordinates.toViewConst();
+  arrayView2d< localIndex > const receiverNodeIds = m_receiverNodeIds.toView();
+  arrayView2d< real64 > const receiverConstants = m_receiverConstants.toView();
+  arrayView1d< localIndex > const receiverIsLocal = m_receiverIsLocal.toView();
+  receiverNodeIds.setValues< serialPolicy >( -1 );
+  receiverConstants.setValues< serialPolicy >( -1 );
+  receiverIsLocal.setValues< serialPolicy >( 0 );
+
+  forTargetRegionsComplete( mesh, [&]( localIndex const,
+                                       localIndex const,
+                                       ElementRegionBase & elemRegion )
+  {
+    elemRegion.forElementSubRegionsIndex< CellElementSubRegion >( [&]( localIndex const,
+                                                                       CellElementSubRegion & elementSubRegion )
+    {
+
+      arrayView2d< localIndex const, cells::NODE_MAP_USD > const & elemsToNodes = elementSubRegion.nodeList();
```

Comment:
Here you may want to check that `numNodesPerElem` is equal to eight, since the assumption is that the acoustic solver is designed for hexahedral meshes. You could add a `GEOSX_THROW_IF` that will check that and terminate the simulation if `elementSubRegion.getElementTypeString() != "C3D8"`

Reply:
If this only works on `C3D8` is the `finiteElement::dispatch3D` below necessary?